### PR TITLE
(maint) test getgrouplist usage

### PR DESCRIPTION
**Pull request by ciprianbadescu**

### CODE REVIEW

**Comment by ciprianbadescu** [comment]
jenkins please test this on redhat7-64a

**Comment by ciprianbadescu** [comment]
jenkins please test this on redhat7-64a

File: lib/puppet/util/posix.rb [hunk 1]
```diff
@@ -1,5 +1,48 @@
 # Utility methods for interacting with POSIX objects; mostly user and group
 module Puppet::Util::POSIX
+  require 'ffi'
+  module GrpLibC
+    extend FFI::Library
+    ffi_lib FFI::Library::LIBC
+    begin
+      attach_function :getgrouplist, [:string, :int, :pointer, :pointer], :int
```
Comment (joshcooper) [inline lib/puppet/util/posix.rb:8]: I think `gid_t` is an unsigned int(?) if so then, the second parameter should be `:uint`

File: lib/puppet/util/posix.rb [hunk 2]
```diff
@@ -1,5 +1,48 @@
...
+      #     ngroups *= 2;
+      #     ngroups_ptr.write_int(ngroups)
+      #   end
+      #   groups_ptr.free if groups_ptr
+      #   groups_ptr = FFI::MemoryPointer.new(:uint, ngroups_ptr.get_int(0))
+      #   ret = GrpLibC.getgrouplist(user, gid, groups_ptr, ngroups_ptr)
+      # end
+      if ret < 0
+        groups_ptr.free if groups_ptr
+        groups_ptr = FFI::MemoryPointer.new(:uint, ngroups_ptr.get_int(0))
```
Comment (joshcooper) [inline lib/puppet/util/posix.rb:36]: MemoryPointer inherits from AbstractMemory, so you can do things like `ngroups_ptr.read_uint` and drop the '0' offset.

File: lib/puppet/util/posix.rb [hunk 3]
```diff
@@ -1,5 +1,48 @@
...
+    rescue FFI::NotFoundError => e
+      puts e.message
+    end
+
+    def self.user_groups(user)
+      gid = Puppet::Etc.getpwnam(user).gid
+      ngroups_ptr = FFI::MemoryPointer.new(:int)
+      ngroups = 16
+      ngroups_ptr.write_int(ngroups)
+      groups_ptr = FFI::MemoryPointer.new(:uint, ngroups)
```
Comment (joshcooper) [inline lib/puppet/util/posix.rb:18]: Ruby will automatically free the memory when the pointer goes out of scope, but I like to make that explicit using the block form

```ruby 
FFI::MemoryPointer.new(:unit, ngroups) { |groups_ptr| ...  }
```

File: lib/puppet/util/posix.rb [hunk 4]
```diff
@@ -1,5 +1,48 @@
 # Utility methods for interacting with POSIX objects; mostly user and group
 module Puppet::Util::POSIX
+  require 'ffi'
+  module GrpLibC
```
Comment (joshcooper) [inline lib/puppet/util/posix.rb:4]: Maybe add this to `lib/puppet/ffi/posix`?

**Pull request closed without merging**